File: hydra-haskell/src/main/haskell/Hydra/Dsl/Lib/Equality.hs
--| Phantom-typed term DSL for the hydra.lib.equality library

module Hydra.Dsl.Lib.Equality where

import Hydra.Core
import Hydra.Mantle
import Hydra.Phantoms
import Hydra.Sources.Libraries
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Dsl.Phantoms

import Data.Int


compare :: TTerm a -> TTerm a -> TTerm Comparison
compare = primitive2 _equality_compare

equal :: TTerm a -> TTerm a -> TTerm Bool
equal = primitive2 _equality_equal

gt :: TTerm a -> TTerm a -> TTerm Bool
gt = primitive2 _equality_gt

gte :: TTerm a -> TTerm a -> TTerm Bool
gte = primitive2 _equality_gte

identity :: TTerm a -> TTerm a
identity = primitive1 _equality_identity

lt :: TTerm a -> TTerm a -> TTerm Bool
lt = primitive2 _equality_lt

lte :: TTerm a -> TTerm a -> TTerm Bool
lte = primitive2 _equality_lte

max :: TTerm a -> TTerm a -> TTerm a
max = primitive2 _equality_max

min :: TTerm a -> TTerm a -> TTerm a
min = primitive2 _equality_min

File: packages/hydra-java/src/main/java/hydra/dsl/meta/lib/Equality.java
package hydra.dsl.meta.lib;
import hydra.dsl.meta.Phantoms;
import hydra.typed.TypedTerm;

/** Phantom-typed term DSL for the {@code hydra.lib.equality} library. */
public final class Equality {
    private Equality() {}

    private static <A> TypedTerm<A> prim(String name) {
        return Phantoms.var("hydra.lib.equality." + name);
    }

    public static <A> TypedTerm<A> compare(TypedTerm<?> x, TypedTerm<?> y) { return Phantoms.apply(prim("compare"), x, y); }
    public static <A> TypedTerm<A> equal(TypedTerm<?> a, TypedTerm<?> b) { return Phantoms.apply(prim("equal"), a, b); }
    public static <A> TypedTerm<A> gt(TypedTerm<?> a, TypedTerm<?> b) { return Phantoms.apply(prim("gt"), a, b); }
    public static <A> TypedTerm<A> gte(TypedTerm<?> a, TypedTerm<?> b) { return Phantoms.apply(prim("gte"), a, b); }
    public static <A> TypedTerm<A> identity(TypedTerm<?> a) { return Phantoms.apply(prim("identity"), a); }
    public static <A> TypedTerm<A> lt(TypedTerm<?> a, TypedTerm<?> b) { return Phantoms.apply(prim("lt"), a, b); }
    public static <A> TypedTerm<A> lte(TypedTerm<?> a, TypedTerm<?> b) { return Phantoms.apply(prim("lte"), a, b); }
    public static <A> TypedTerm<A> max_(TypedTerm<?> a, TypedTerm<?> b) { return Phantoms.apply(prim("max"), a, b); }
    public static <A> TypedTerm<A> min_(TypedTerm<?> a, TypedTerm<?> b) { return Phantoms.apply(prim("min"), a, b); }
}
